bobtimus/src/coin_selection.rs

```diff
@@ -0,0 +1,259 @@
+use anyhow::Result;
+use bdk::{
+    bitcoin::Amount,
+    database::{BatchOperations, Database},
+    wallet::coin_selection::CoinSelectionResult,
+    wallet::coin_selection::{BranchAndBoundCoinSelection, CoinSelectionAlgorithm},
+};
+use elements_fun::{ExplicitTxOut, OutPoint};
+
+#[derive(Clone, PartialEq)]
+pub struct Utxo {
+    outpoint: OutPoint,
+    txout: ExplicitTxOut,
+}
+
+impl From<Utxo> for bdk::UTXO {
+    fn from(utxo: Utxo) -> Self {
+        let value = utxo.txout.value.0;
+        let script_pubkey = utxo.txout.script_pubkey.into_bytes();
+        let script_pubkey = bdk::bitcoin::Script::from(script_pubkey);
+
+        Self {
+            outpoint: bdk::bitcoin::OutPoint {
+                txid: bdk::bitcoin::Txid::from_hash(utxo.outpoint.txid.as_hash()),
+                vout: utxo.outpoint.vout,
+            },
+            txout: bdk::bitcoin::TxOut {
+                value,
+                script_pubkey,
+            },
+            script_type: bdk::ScriptType::External,
+        }
+    }
+}
+
+pub fn coin_select(utxos: Vec<Utxo>, target: Amount) -> Result<Vec<Utxo>> {
```

Comment:
We will need the fee as an output as well, plus the change amount.

Reply:
We can output the recommended fee, but it currently only takes into account the weight of each input and _not_ the weight of whatever will be added to the transaction to satisfy it. Meaning it is an unreliable estimate. Do you think it would be sufficient to assume the highest possible "satisfaction weight"? Or is it worth trying to figure out the value from the script, a bit like miniscript does (maybe there's code for this in bitcoind)?

The change amount should just be the result of adding all the selected UTXOs, minus the estimated fee, but we can return it explicitly if it's preferable.

Reply:
> The change amount should just be the result of adding all the selected UTXOs, minus the estimated fee, but we can return it explicitly if it's preferable.

Fair. I agree that this makes the return values not completely orthogonal but at least in elements, we need to be explicit about the fee which means every caller will have to do this summing.

> We can output the recommended fee, but it currently only takes into account the weight of each input and not the weight of whatever will be added to the transaction to satisfy it. Meaning it is an unreliable estimate. Do you think it would be sufficient to assume the highest possible "satisfaction weight"? Or is it worth trying to figure out the value from the script, a bit like miniscript does (maybe there's code for this in bitcoind)?

At least in the browser-based wallet, we can assume a fixed weight per UTXO because it is always the same script (p2wpkh).

Reply:
We only have standard scripts on both sides, so a fixed value per script pub key should be fine. It is obviously not a general solution (that would require miniscript) but it should be good enough for our usecase.

`Script` has `is_XYZ` classifiers which you should be able to use to decide on a reasonable estimate.

Reply:
Let's hard-code it to the max satisfaction weight of a p2wpkh script then, because we don't need this to be a general purpose coin selection algorithm yet anyway.

> Script has is_XYZ classifiers which you should be able to use to decide on a reasonable estimate.

I've been trying to use them, but some of the variants don't have obvious (to me) max satisfaction weights.

Reply:
> We only have standard scripts on both sides, so a fixed value per script pub key should be fine. It is obviously not a general solution (that would require miniscript) but it should be good enough for our usecase.

The elements-miniscript fork seems to make some progress: https://github.com/sanket1729/elements-miniscript. I don't know if it's usable yet though.

For now we can assume a fixed satisfaction weight because we only support `P2WPK` anyways.

Reply:
> Let's hard-code it to the max satisfaction weight of a p2wpkh script then, because we don't need this to be a general purpose coin selection algorithm yet anyway.
> 
> > Script has is_XYZ classifiers which you should be able to use to decide on a reasonable estimate.
> 
> I've been trying to use them, but some of the variants don't have obvious (to me) max satisfaction weights.

You don't need to handle all of the variants. Witness and non-witness p2pk should be fine. Any other utxos (on bobtimus' side) can be filtered out!

Reply:
> The elements-miniscript fork seems to make some progress: https://github.com/sanket1729/elements-miniscript. I don't know if it's usable yet though.

That looked promising when I took a look, but the problem that I see is that the UTXOs that we get from esplora or elementsd don't have a descriptor associated with them (AFAIK).

Reply:
> You don't need to handle all of the variants. Witness and non-witness p2pk should be fine. Any other utxos (on bobtimus' side) can be filtered out!

I like that approach: handle the two relevant cases and filter out the rest of the provided UTXOs that don't apply.